src/NUnitFramework/framework/Constraints/ThrowsConstraint.cs

```diff
@@ -197,22 +196,36 @@ private static IInvocationDescriptor GetInvocationDescriptor(object actual)
                 {
                     var testDelegate = actual as TestDelegate;
 
-                    if (testDelegate == null)
-                        throw new ArgumentException(
-                            String.Format("The actual value must be a TestDelegate or ActualValueDelegate but was {0}",
-                                actual.GetType().Name),
-                            "actual");
```

Comment:
I had to restructure the flow in order to support both TestDelegate and AsyncTestDelegate. Unless I made a mistake, the overall logic should be the same as before.